authentication/azure_sp_objectid.go

```diff
@@ -3,47 +3,144 @@ package authentication
 import (
 	"context"
 	"fmt"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"strings"
 
 	"github.com/Azure/azure-sdk-for-go/services/graphrbac/1.6/graphrbac"
+	"github.com/manicminer/hamilton/auth"
+	"github.com/manicminer/hamilton/environments"
+	"github.com/manicminer/hamilton/msgraph"
+	"github.com/manicminer/hamilton/odata"
+
 	"github.com/hashicorp/go-azure-helpers/sender"
 )
 
 func buildServicePrincipalObjectIDFunc(c *Config) func(ctx context.Context) (string, error) {
 	return func(ctx context.Context) (string, error) {
-		env, err := AzureEnvironmentByNameFromEndpoint(ctx, c.MetadataHost, c.Environment)
-		if err != nil {
+		objectId, err := objectIdFromMsGraph(ctx, c)
+		if err == nil {
+			return objectId, nil
+		}
+		if !strings.HasPrefix(err.Error(), "access denied") {
 			return "", err
 		}
 
-		s := sender.BuildSender("GoAzureHelpers")
+		log.Printf("Unable to retrieve service principal object ID from Microsoft Graph: %+v", err)
+		return objectIdFromAadGraph(ctx, c)
```

Comment:
we should have the context on the `config` here to know which to use, so perhaps we don't need this fallback here?

Reply:
I tried to separate the concerns of selecting a token version and choosing which API to call, given that both token versions work with both APIs and there are different prerequisites.

Having it gracefully fall back is the first pass - I'm hoping to find a neat way of exposing this downstream so that a deprecation notice can be raised. WDYT?

Reply:
I think I misunderstood what you meant - we could make this configurable instead of having auto fallback. Actually now I'm thinking this would be better since it can be explicitly opted into, and when it fails it's directly due to user configuration which feels more obvious/transparent.